great_expectations/datasource/fluent/sources.py

```diff
@@ -333,31 +334,101 @@ def pandas_default(self) -> PandasDatasource:
                 ) from e
-            old_datasource = None
-        if old_datasource and not isinstance(old_datasource, datasource_type):
+            current_datasource = None
+        if current_datasource and not isinstance(current_datasource, datasource_type):
             raise ValueError(
                 f"Trying to update datasource {name} but it is not the correct type. "
-                f"Expected {datasource_type} but got {type(old_datasource)}"
+                f"Expected {datasource_type} but got {type(current_datasource)}"
             )
 
+    def _datasource_passed_in_as_only_argument(
+        self, datasource_type, first: Optional[str, Datasource], **kwargs
+    ) -> Optional[Datasource]:
+        """Returns a datasource if one is passed in, otherwise None."""
+        from great_expectations.datasource.fluent.interfaces import Datasource
+
+        datasource: Optional[Datasource] = None
+        if first and isinstance(first, Datasource):
+            if len(kwargs) != 0:
+                raise ValueError(
+                    f"The datasource must be the sole argument. We also received: {kwargs}"
+                )
+            datasource = first
+        elif first is None and "datasource" in kwargs:
+            if len(kwargs) != 1:
+                raise ValueError(
+                    f"The datasource must be the sole argument. We received: {kwargs}"
+                )
+            datasource = kwargs["datasource"]
+        if datasource and not isinstance(datasource, datasource_type):
+            raise ValueError(
+                f"Trying to modify datasource {datasource.name} but it is not the correct type. "
+                f"Expected {datasource_type} but got {type(datasource)}"
+            )
+        return datasource
+
+    def _datasource_passed_in(
+        self, datasource_type, first: Optional[str, Datasource], **kwargs
+    ) -> Optional[Datasource]:
+        """Validates the input is a datasource or a set of constructor parameters
+
+        The first argument can be a non-keyword argument. If present it is a datasource
+        or the name. If it is None, we expect to find either datasource or name as a kwarg.
+
+        Args:
+             datasource_type: The expected type of datasource
+             first: Either the datasource or the name of the datasource.
+
+        Returns:
+            The passed in datasource or None.
+
+        Raises:
+            ValueError: This is raised if a datasource is passed in with additional arguments or
+                        a datasource is not passed in and no name argument is present.
+        """
+        new_datasource = self._datasource_passed_in_as_only_argument(
+            datasource_type, first, **kwargs
+        )
+        if new_datasource:
+            return new_datasource
+        if (first and isinstance(first, str) and "name" not in "kwargs") or (
+            first is None and "name" in kwargs and isinstance(kwargs["name"], str)
+        ):
```

Comment:
@billdirks it just occurred to me... this might actually be a good case for `single-dispatch`.
Dispatching on the value of `first` being either `str` or a `Datasource`.

https://docs.python.org/3/library/functools.html#functools.singledispatch
https://rednafi.github.io/digressions/python/2020/04/05/python-singledispatch.html

Reply:
You are right, this is a prime candidate for single dispatch. Depending on time, I may or may not update this for this PR but will definitely have this more on my radar.

Reply:
I'm resolving this conversation since github keeps opening a new comment box here and it's making the UI unusable.